crypto/pkcs7/pk7_doit.c

```diff
@@ -64,20 +64,29 @@ static int pkcs7_bio_add_digest(BIO **pbio, X509_ALGOR *alg,
     BIO *btmp;
     const char *name;
     EVP_MD *fetched = NULL;
+    const EVP_MD *md;
 
     if ((btmp = BIO_new(BIO_f_md())) == NULL) {
         PKCS7err(PKCS7_F_PKCS7_BIO_ADD_DIGEST, ERR_R_BIO_LIB);
         goto err;
     }
 
     name = OBJ_nid2sn(OBJ_obj2nid(alg->algorithm));
+
+    (void)ERR_set_mark();
     fetched = EVP_MD_fetch(ctx->libctx, name, ctx->propq);
-    if (fetched == NULL) {
+    if (fetched != NULL)
+        md = fetched;
+    else
+        md = EVP_get_digestbyname(name);
+
+    if (md == NULL) {
         PKCS7err(PKCS7_F_PKCS7_BIO_ADD_DIGEST, PKCS7_R_UNKNOWN_DIGEST_TYPE);
```

Comment:
You should also clear the ERR mark here.

Reply:
Fixed. Thanks!